### DISCUSSION TITLE
Tools or methods for testing project performance？

### DISCUSSION THREAD

**Discussion by Bin-ly** [sharkdp/fd#1451] [General]
Hello, I have recently been investigating some projects to rewrite CLI tools in rust, looking for specific advantages of rust compared to c. Here are a few questions for you:
1. What is your original intention for this project? What are the advantages of comparing C's find tool? (Because rust has always emphasized security, does fd improve memory safety compared to find?)
2. Do you use any tools or methods to check the performance of the code? (e.g. robustness, security, presence of errors and bugs)

**Comment by tavianator** [comment]
I am not @sharkdp, but here's my two cents:

> What are the advantages of comparing C's find tool? (Because rust has always emphasized security, does fd improve memory safety compared to find?)

I would say that Rust's "fearless concurrency" is the main benefit.  By running in parallel, `fd` runs much faster than typical `find` implementations.  Parallelizing an existing `find` implementation written in C is very challenging to do correctly and efficiently, which I know because I have [done it](https://tavianator.com/2023/bfs_3.0.html)).

> Do you use any tools or methods to check the performance of the code? (e.g. robustness, security, presence of errors and bugs)

`fd` has a set of [benchmarks](https://github.com/sharkdp/fd-benchmarks) to assess (runtime) performance.  It also has a test suite and CI infrastructure to catch errors and bugs, particularly across platforms.

There are occasional security-relevant issues in the design and operation of `fd`, such as #760.  From what I know these are analyzed and considered case-by-case.  I don't believe we use any automated tools or methods for security analysis, other than standard compiler analysis and linting (rustc + Clippy, basically).  That also overlaps with "errors and bugs" above, I guess.

The situation is similar for robustness.  We have some relevant issues like #288, #1186, which are also handled case-by-case.  We do add regression tests for things like this when feasible.  I don't think we use any automated tools or methods for this either.  It might be interesting to try fuzzing, but that would be harder for `fd` than most other CLI tools with a more limited input/output structure.
